simtools/layout/geo_coordinates.py

```diff
@@ -0,0 +1,163 @@
+    """
+    Geospatial Coordinate systems.
+    Defines UTM, WGS84 and local (CORSIKA) coordinate systems.
+
+    """
+
+    def __init__(self):
+        """
+        Initialize GeoCoordinates
+
+        """
+
+        self._logger = logging.getLogger(__name__)
+
+    def crs_utm(self, epsg):
+        """
+        UTM coordinate system definition.
+
+        Parameters
+        ----------
+        epsg: int
+            EPSG code for UTM zone.
+
+        Returns
+        -------
+        pyproj.CRS
+            UTM coordinate system.
+
+        """
+        if epsg:
+            crs_utm = pyproj.CRS.from_user_input(epsg)
+            self._logger.debug(f"UTM coordinate system: {crs_utm}")
+            return crs_utm
+
+        return None
+
+    @staticmethod
+    def crs_wgs84():
+        """
+        WGS84 coordinate system definition.
+
+        Returns
+        -------
+        pyproj.CRS
+            WGS84 coordinate system.
+
+        """
+        return pyproj.CRS("EPSG:4326")
+
+    def crs_local(self, reference_point):
+        """
+        Local coordinate system definition.
+
+        Parameters
+        ----------
+        reference_point: simtools.layout.telescope_position
+            Reference coordinate.
+
+        Returns
+        -------
+        pyproj.CRS
+            local coordinate system.
+
+        """
+        try:
+            _center_lat, _center_lon, _ = reference_point.get_coordinates("mercator")
+            _scale_factor_k_0 = self._coordinate_scale_factor(reference_point)
+            if not np.isnan(_center_lat.value) and not np.isnan(_center_lon.value):
+                proj4_string = (
+                    "+proj=tmerc +ellps=WGS84 +datum=WGS84"
+                    f" +lon_0={_center_lon} +lat_0={_center_lat}"
+                    f" +axis=nwu +units=m +k_0={_scale_factor_k_0}"
+                )
+                crs_local = pyproj.CRS.from_proj4(proj4_string)
+                self._logger.debug(f"Local (CORSIKA) coordinate system: {crs_local}")
+                return crs_local
+        except AttributeError:
+            self._logger.error("Failed to derive local coordinate system. Missing reference point")
+            raise
+
+        return None
+
+    def _coordinate_scale_factor(self, reference_point):
+        """
+        Derive coordinate scale factor for transformation into
+        local coordinate system
+
+        Depends on latitude and altitude of array center.
+        Ignores transformation of geodetic height to geocentric height
+        (this introduces an error on the sub-mm scale).
+
+        Parameters
+        ----------
+        reference_point: simtools.layout.telescope_position
+            Reference coordinate.
+
+        Returns
+        -------
+        k_0: float
+            Scale factor for local coordinate system.
+
+        """
+
+        try:
+            _center_lat, _, _centre_altitude = reference_point.get_coordinates("mercator")
+        except TypeError:
```

Comment:
Now I am a bit confused. In the previous function you tried and excepted an `AttributeError`. Now the exception is a `TypeError`. Is the `try` and `except` in the previous function specifically aimed to catch the `AttributeError` raised in `_coordinate_scale_factor` (below)? If so, why is the try..except around the entire block instead of just specifically this function?

Reply:
Thanks for the comment, this was important:

- raise now an AttributeError when reference point is None
- remove the second try..except, as we are pulling the WGS84 Geodetics from pyproj (depends on no input parameters and therefore should always work)